File: scripts/fixIslandData.js
/**
 * fixIslandData.js
 * 問題1: 離島の不正な travelTime を修正（日帰り不可能な離島が表示される原因）
 * 問題2: 離島のじゃらんキーワードが本島都市になっている箇所を島名に修正
 */
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const DEST_FILE = path.join(__dirname, '../src/data/destinations.json');
const data = JSON.parse(fs.readFileSync(DEST_FILE, 'utf-8'));

const JALAN_AFF = 'https://ck.jp.ap.valuecommerce.com/servlet/referral?sid=3764408&pid=892559858&vc_url=';

/** Shift-JIS バイト列を percent-encode してじゃらん affiliate URL を生成 */
function buildJalanUrl(shiftjisHex) {
  // shiftjisHex: 例 '92BC93C5' (直島)
  const bytes = shiftjisHex.match(/.{2}/g);
  const kwEncoded = bytes.map(b => `%${b}`).join(''); // %92%BC%93%C5
  const jalanInner = `https://www.jalan.net/uw/uwp2011/uww2011init.do?keyword=${kwEncoded}`;
  // vc_url 内でさらに percent-encode （% → %25）
  const encoded = jalanInner
    .replace(/:/g,  '%3A')
    .replace(/\//g, '%2F')
    .replace(/\?/g, '%3F')
    .replace(/=/g,  '%3D')
    .replace(/%/g,  '%25');
  // ただし既に %25 にした後の %3A 等がある → 正規化済み
  // 実際の手順: jalanInner をまるごと encodeURIComponent と同等に処理
  const vc = encodeURIComponent(jalanInner).replace(/%25/g, '%25'); // already percent-encoded
  return `${JALAN_AFF}${encodeURIComponent(jalanInner)}`;
}

// ─── 問題1: travelTime 修正テーブル ────────────────────────────────────────
// { id, city, oldVal, newVal, reason }
const TRAVEL_TIME_FIXES = [
  // 瀬戸内海
  { id: 'shodoshima', city: 'takamatsu', old: 5,   val: 65,  reason: '高松港→土庄港 フェリー最速35分＋港アクセス30分' },
  { id: 'naoshima',   city: 'takamatsu', old: 5,   val: 50,  reason: '高松港→宮浦港 高速艇25分＋港アクセス25分' },
  { id: 'teshima',    city: 'takamatsu', old: 15,  val: 40,  reason: '高松港→豊島家浦港 フェリー約35分' },
  // 伊豆諸島
  { id: 'kozushima',  city: 'tokyo',     old: 20,  val: 145, reason: '竹芝→神津島 ジェット船約2時間20分＋竹芝まで5分' },
  { id: 'izu-oshima', city: 'tokyo',     old: 20,  val: 115, reason: '竹芝→大島 ジェット船約1時間45分＋竹芝まで30分' },
  // 九州・長崎
  { id: 'goto-islands', city: 'fukuoka', old: 20,  val: 90,  reason: '博多→福江 高速船約85分' },
  { id: 'tsushima',   city: 'fukuoka',   old: 20,  val: 50,  reason: '福岡空港→対馬空港 空路約40分＋空港アクセス10分' },
  { id: 'iki',        city: 'fukuoka',   old: 20,  val: 75,  reason: '博多港→芦辺港 ジェット船65分＋港アクセス10分' },
  // 沖縄・離島
  { id: 'ejima',      city: 'naha',      old: 30,  val: 100, reason: '那覇→本部バス70分＋本部→伊江島フェリー30分' },
  // 新潟
  { id: 'sado-island', city: 'tokyo',    old: 110, val: 170, reason: '東京→新潟新幹線95分＋ジェットフォイル67分＋乗換8分' },
];

// ─── 問題2: じゃらんキーワード修正テーブル ────────────────────────────────
// Shift-JIS hex は Python: keyword.encode('shift_jis').hex().upper() で生成
const JALAN_KW_FIXES = [
  { id: 'mihogaseki',       kw: '美保関',     hex: '8CA786CA8AD4' },
  { id: 'amakusa',          kw: '天草',       hex: '93V91AC' },      // 天=9356, 草=8FA3... let me compute properly below
  { id: 'wajima',           kw: '輪島',       hex: '97D68C87' },
  { id: 'shakotan',         kw: '積丹',       hex: '90CF924A' },
  { id: 'itoshima',         kw: '糸島',       hex: '88CF93C5' },
  { id: 'shimabara',        kw: '島原',       hex: '93C58C9F' },
  { id: 'shishijima',       kw: '志々島',     hex: '8E54929B93C5' },
  { id: 'minami-izu',       kw: '南伊豆',     hex: '8CF48CE090B9' },
  { id: 'nakijin',          kw: '今帰仁',     hex: '8DA662A48CD1' },
  { id: 'motobu',           kw: '本部',       hex: '96{'+'{'+'90'+'7B' },  // placeholder, recomputed below
  { id: 'onna',             kw: '恩納村',     hex: '89BA945892BA' },
  { id: 'hateruma-island',  kw: '波照間',     hex: '944E9253924C' },
  { id: 'kakeromajima',     kw: '加計呂麻',   hex: '89C597B98CE48CBC' },
  { id: 'tobushima',        kw: '飛島',       hex: '94F293C5' },
  { id: 'awashima',         kw: '粟島',       hex: '88F793C5' },
  { id: 'tarama-island',    kw: '多良間島',   hex: '91BD97C08CDC93C5' },
  { id: 'mishima-yamaguchi',kw: '見島',       hex: '8CA593C5' },
  { id: 'oshima-ehime',     kw: '大三島',     hex: '91E88EO93C5' }, // placeholder
  { id: 'ibuki-island',     kw: '伊吹島',     hex: '88C293C593C5' }, // placeholder
  { id: 'ogijima',          kw: '男木島',     hex: '92j93C5' }, // placeholder
  { id: 'megijima',         kw: '女木島',     hex: '8F7793C593C5' }, // placeholder
  { id: 'motoshima',        kw: '本島',       hex: '96{7093C5' }, // placeholder
  { id: 'sanagijima',       kw: '佐柳島',     hex: '8D645F93C5' }, // placeholder
  { id: 'awashima-kagawa',  kw: '粟島',       hex: '88F793C5' },
  { id: 'sensui-island',    kw: '仙酔島',     hex: '90969154B93C5' }, // placeholder
  { id: 'manabe-island',    kw: '真鍋島',     hex: '9088925793C5' }, // placeholder
];

// ─────────────────────────────────────────────────────────────────────────────
// Hex テーブルは Python で事前計算済みのもので置き換える
// ─────────────────────────────────────────────────────────────────────────────

const changes = [];

// --- 問題1: travelTime 修正 ---
for (const fix of TRAVEL_TIME_FIXES) {
  const dest = data.find(d => d.id === fix.id);
  if (!dest) { console.warn(`⚠️  ID not found: ${fix.id}`); continue; }
  if (!dest.travelTime) continue;
  const current = dest.travelTime[fix.city];
  if (current === undefined) { console.warn(`⚠️  city ${fix.city} not in ${fix.id}`); continue; }
  dest.travelTime[fix.city] = fix.val;
  changes.push({
    id: fix.id, name: dest.name,
    type: 'travelTime',
    detail: `${fix.city}: ${current} → ${fix.val}分 (${fix.reason})`,
  });
}

// --- 問題2はPythonで実行 ---
// (Shift-JISエンコードはNode.jsで直接扱えないためPython側で処理)

fs.writeFileSync(DEST_FILE, JSON.stringify(data, null, 2));

console.log('\n=== travelTime 修正完了 ===');
for (const c of changes) {
  console.log(`✓ ${c.name} (${c.id}): ${c.detail}`);
}
console.log(`\n合計: ${changes.length}件修正`);

File: scripts/fixIslandData.py
#!/usr/bin/env python3
"""
fixIslandData.py
問題1: 離島の不正な travelTime を修正
問題2: じゃらんキーワードが本島都市の離島を島名キーワードに修正
"""
import json, re
from urllib.parse import quote, unquote
from pathlib import Path

DEST_FILE = Path(__file__).parent.parent / 'src/data/destinations.json'
JALAN_AFF = 'https://ck.jp.ap.valuecommerce.com/servlet/referral?sid=3764408&pid=892559858&vc_url='

data = json.loads(DEST_FILE.read_text('utf-8'))

def build_jalan_url(keyword: str) -> str:
    """キーワード（日本語）からじゃらんアフィリエイトURLを生成"""
    kw_bytes = keyword.encode('shift_jis')
    kw_pct = ''.join(f'%{b:02X}' for b in kw_bytes)
    jalan_inner = f'https://www.jalan.net/uw/uwp2011/uww2011init.do?keyword={kw_pct}'
    # vc_url パラメータ用にエンコード（%→%25 を含む）
    return JALAN_AFF + quote(jalan_inner, safe='')

def decode_jalan_kw(url: str) -> str:
    """現在のじゃらんURLからキーワードをShift-JISデコード"""
    if 'vc_url=' not in url:
        return ''
    vc_decoded = unquote(url[url.find('vc_url=')+7:])
    m = re.search(r'keyword=([^&]+)', vc_decoded)
    if not m:
        return ''
    kw_bytes = bytes(int(x, 16) for x in re.findall(r'%([0-9A-Fa-f]{2})', m.group(1)))
    try:
        return kw_bytes.decode('shift_jis')
    except Exception:
        return ''

# ════════════════════════════════════════════════════
# 問題1: travelTime 修正
# ════════════════════════════════════════════════════
TRAVEL_FIXES = [
    # (id, city, expected_old, new_val, reason)
    ('shodoshima',    'takamatsu', 5,   65,  '高松港→土庄港 フェリー最速35分＋港アクセス30分'),
    ('naoshima',      'takamatsu', 5,   50,  '高松港→宮浦港 高速艇25分＋港アクセス25分'),
    ('teshima',       'takamatsu', 15,  40,  '高松港→豊島家浦港 フェリー約35分'),
    ('kouzushima',    'tokyo',     20, 145,  '竹芝→神津島 ジェット船約2時間20分＋竹芝まで5分'),
    ('izu-oshima',    'tokyo',     20, 115,  '竹芝→大島 ジェット船約1時間45分＋竹芝まで30分'),
    ('goto',          'fukuoka',   20,  90,  '博多→福江 高速船約85分'),
    ('tsushima',      'fukuoka',   20,  50,  '福岡空港→対馬空港 空路約40分＋空港アクセス10分'),
    ('iki-island',    'fukuoka',   20,  75,  '博多港→芦辺港 ジェット船65分＋港アクセス10分'),
    ('ie-island',     'naha',      30, 100,  '那覇→本部バス70分＋本部→伊江島フェリー30分'),
    ('sado-island',   'tokyo',    110, 170,  '東京→新潟新幹線95分＋ジェットフォイル67分＋乗換8分'),
]

tt_changes = []
for dest_id, city, old_val, new_val, reason in TRAVEL_FIXES:
    dest = next((d for d in data if d['id'] == dest_id), None)
    if not dest:
        print(f'⚠️  ID未発見: {dest_id}')
        continue
    tt = dest.get('travelTime', {})
    cur = tt.get(city)
    if cur is None:
        print(f'⚠️  {dest["name"]}: city "{city}" なし')
        continue
    if cur != old_val:
        print(f'⚠️  {dest["name"]} {city}: 期待値={old_val}, 実際={cur} → そのまま {new_val} に修正')
    tt[city] = new_val
    tt_changes.append((dest['name'], city, cur, new_val, reason))

# ════════════════════════════════════════════════════
# 問題2: じゃらんキーワード修正
# ════════════════════════════════════════════════════
# (id, 正しいキーワード)
JALAN_FIXES = [
    ('mihonoseki',        '美保関'),
    ('amakusa',           '天草'),
    ('wajima',            '輪島'),
    ('shakotan',          '積丹'),
    ('itoshima',          '糸島'),
    ('shimabara',         '島原'),
    ('shijishima',        '志々島'),
    ('minami-izu',        '南伊豆'),
    ('nakijin',           '今帰仁'),
    ('motobu',            '本部'),
    ('onna',              '恩納'),
    ('hateruma-island',   '波照間島'),
    ('kakeroma-island',   '加計呂麻島'),
    ('tobishima-island',  '飛島'),
    ('awashima-island',   '粟島'),
    ('tarama-island',     '多良間島'),
    ('mishima-yamaguchi', '見島'),
    ('omishima-island',   '大三島'),
    ('ibukijima',         '伊吹島'),
    ('ogijima',           '男木島'),
    ('megijima',          '女木島'),
    ('honjima',           '本島'),
    ('sanagijima',        '佐柳島'),
    ('awashima-kagawa',   '粟島'),
    ('sensui-jima',       '仙酔島'),
    ('manabeshima',       '真鍋島'),
]

jalan_changes = []
for dest_id, new_kw in JALAN_FIXES:
    dest = next((d for d in data if d['id'] == dest_id), None)
    if not dest:
        print(f'⚠️  ID未発見: {dest_id}')
        continue
    hl = dest.get('hotelLinks', {})
    jal = hl.get('jalan', '')
    if not jal:
        print(f'⚠️  {dest["name"]}: じゃらんURL未設定')
        continue
    old_kw = decode_jalan_kw(jal)
    if old_kw == new_kw:
        print(f'  ✓ {dest["name"]}: すでに正しい ({new_kw})')
        continue
    new_url = build_jalan_url(new_kw)
    hl['jalan'] = new_url
    jalan_changes.append((dest['name'], old_kw, new_kw))

# ════════════════════════════════════════════════════
# 保存 & レポート
# ════════════════════════════════════════════════════
DEST_FILE.write_text(json.dumps(data, ensure_ascii=False, indent=2), 'utf-8')

print('\n' + '='*60)
print('【問題1: travelTime 修正】')
print('='*60)
for name, city, old, new, reason in tt_changes:
    print(f'  {name} / {city}: {old}分 → {new}分')
    print(f'      理由: {reason}')

print(f'\n  合計: {len(tt_changes)} 件\n')

print('='*60)
print('【問題2: じゃらんキーワード修正】')
print('='*60)
for name, old, new in jalan_changes:
    print(f'  {name}: 「{old}」 → 「{new}」')

print(f'\n  合計: {len(jalan_changes)} 件')
print('\n✅ destinations.json 更新完了')
